### PR TITLE
Espnet-SE, Speech enhancement recipes

### PR DESCRIPTION
**Pull request by LiChenda**
This is the second pull request split from #2649 . In this PR, we mainly include some recipes that we have prepared in the last few months. Here are the updates:

New recipes:

  - WHAMR/enh1
  - WHAM/enh1
  - ChiME4/enh1
  - sms_wsj/enh1
  - librimix/enh1
  
Updated recipes :

  - wsj0_2mix/enh1
  - wsj0_2mix_spatialized/enh1

Others:
  - Update some readmefile
  - Upddate some TEMPLATE recipe related scripts
  - Bug fixs

Thanks to @Emrys365 for working on the recipes together.

### CODE REVIEW

**Comment by sw005320** [comment]
Pre-trained models?
There are some discussions in https://github.com/espnet/espnet/pull/2878

**Comment by sw005320** [comment]
For the pre-trained model preparation, maybe we can distribute the effort to the others after this PR is merged.
This would be a good check of each recipe.

**Comment by LiChenda** [comment]
> Pre-trained models?
> There are some discussions in #2878

Since we have refactor the enhancement models in to new structure, the previous trained models are not compatible with the new codes. We are training new models for these recipes and thus the pretrained models are not included here. Once they are done, I'll upload the models. I think they shoud be done in this PR ? @sw005320

File: egs2/librimix/enh1/local/data.sh [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+. utils/parse_options.sh
+
+if [ $# -ne 0 ]; then
+    echo "${help_message}"
+    exit 1;
+fi
+
+# train_set="tr_"${min_or_max}_${sample_rate}
+# train_dev="cv_"${min_or_max}_${sample_rate}
+# recog_set="tt_"${min_or_max}_${sample_rate}
```
Comment (kan-bayashi) [inline egs2/librimix/enh1/local/data.sh:37]: Remove?

File: egs2/librimix/enh1/local/data.sh [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  metadata_dir=metadata/Libri$n_src"Mix"
+  python scripts/create_librimix_from_metadata.py --librispeech_dir $LIBRISPEECH \
+    --wham_dir ${cdir}/data/wham_noise \
+    --metadata_dir $metadata_dir \
+    --librimix_outdir $librimix_outdir \
+    --n_src $n_src \
+    --freqs 8k 16k \
+    --modes min max \
+    --types mix_clean mix_both mix_single
+    # --types mix_single
```
Comment (kan-bayashi) [inline egs2/librimix/enh1/local/data.sh:67]: Remove?

File: egs2/librimix/enh1/local/data.sh [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+# recog_set="tt_"${min_or_max}_${sample_rate}
+
+cdir=$PWD
+
+
+git clone https://github.com/JorisCos/LibriMix ./data/LibriMix
+
+
+# the simulation program will write data to wham_noie, so copy it to user directory in case of permission issues
+rsync -r -P ${WHAM} ${cdir}/data/wham_noise
```
Comment (kan-bayashi) [inline egs2/librimix/enh1/local/data.sh:46]: Maybe it is better to add the check of `${WHAM}` and `${LIBRISPEECH}`?

File: egs2/TEMPLATE/enh1/README.md [hunk 4]
```diff
@@ -1,28 +1,44 @@
...
-This is the common recipe for for ESPnet2 speech enhancement frontend. Currently, two speech separation recipes (`wsj0_2mix` and `wsj0_2mix_spatialized`) are supported:
+This is the common recipe for for ESPnet2 speech enhancement frontend. Currently, ten speech enhancement/separation recipes are supported:
 ```
 egs2/
+├── chime4
+│   └── enh1
+├── dirha_wsj
+│   └── enh1
+├── libri_css
+│   └── enh1
```
Comment (kan-bayashi) [inline egs2/TEMPLATE/enh1/README.md:11]: Does not exist?

File: egs2/whamr/enh1/local/whamr_create_mixture.sh [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+echo "Downloading WHAMR! data generation scripts and documentation."
+
+if [ -z "$wham_noise" ]; then
+  # 17.65 GB unzipping to 35 GB
+  wham_noise_url=https://storage.googleapis.com/whisper-public/wham_noise.zip
+  wget --continue -O $wdir/wham_noise.zip ${wham_noise_url}
+  if [ $(ls ${dir}/wham_noise 2>/dev/null | wc -l) -eq 4 ]; then
+    echo "'${dir}/wham_noise/' already exists. Skipping..."
+  else
+    unzip ${wdir}/wham_noise.tar.gz -d ${dir}
```
Comment (kan-bayashi) [inline egs2/whamr/enh1/local/whamr_create_mixture.sh:48]:
```suggestion
    unzip ${wdir}/wham_noise.zip -d ${dir}
```

File: egs2/whamr/enh1/local/data.sh [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+#   - `clean`: a mixture of speech1 and speech2 (for speech separation)
+#   - `single`: a mixture of speech1 and noise (for speech enhancement)
+local/whamr_data_prep.sh --min-or-max ${min_or_max} --sample-rate ${sample_rate} \
+    ${whamr_scripts}/whamr_scripts ${whamr_wav} ${wsj_full_wav} || exit 1;
+
+
+### Also need wsj corpus to prepare language information
+### This is from Kaldi WSJ recipe
+log "local/wsj_data_prep.sh ${WSJ0}/??-{?,??}.? ${WSJ1}/??-{?,??}.?"
+local/wsj_data_prep.sh ${WSJ0}/??-{?,??}.? ${WSJ1}/??-{?,??}.?
```
Comment (kan-bayashi) [inline egs2/whamr/enh1/local/data.sh:70]: `local/wsj_data_prep.sh` is missing.

File: egs2/whamr/enh1/local/data.sh [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+log "local/wsj_format_data.sh"
+local/wsj_format_data.sh
```
Comment (kan-bayashi) [inline egs2/whamr/enh1/local/data.sh:72]: `local/wsj_format_data.sh` is missing.

File: egs2/wham/enh1/local/data.sh [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+#   - `single`: a mixture of speech1 and noise (for speech enhancement)
+local/wham_data_prep.sh --min-or-max ${min_or_max} --sample-rate ${sample_rate} \
+    ${wham_scripts}/wham_scripts ${wham_wav} ${wsj_full_wav} || exit 1;
+
+
+
+### Also need wsj corpus to prepare language information
+### This is from Kaldi WSJ recipe
+log "local/wsj_data_prep.sh ${WSJ0}/??-{?,??}.? ${WSJ1}/??-{?,??}.?"
+local/wsj_data_prep.sh ${WSJ0}/??-{?,??}.? ${WSJ1}/??-{?,??}.?
```
Comment (kan-bayashi) [inline egs2/wham/enh1/local/data.sh:69]: missing.

File: egs2/wham/enh1/local/data.sh [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+log "local/wsj_format_data.sh"
+local/wsj_format_data.sh
```
Comment (kan-bayashi) [inline egs2/wham/enh1/local/data.sh:71]: missing.

File: egs2/whamr/enh1/local/data.sh (hunk 6 above)
Comment (kan-bayashi) [reply egs2/whamr/enh1/local/data.sh:70]: Related files are also needed.

File: egs2/wham/enh1/local/data.sh (hunk 9 above)
Comment (Emrys365) [reply egs2/wham/enh1/local/data.sh:71]: done

File: egs2/wham/enh1/local/data.sh (hunk 8 above)
Comment (Emrys365) [reply egs2/wham/enh1/local/data.sh:69]: done

File: egs2/whamr/enh1/local/data.sh (hunk 7 above)
Comment (Emrys365) [reply egs2/whamr/enh1/local/data.sh:75]: done

File: egs2/whamr/enh1/local/data.sh (hunk 6 above)
Comment (Emrys365) [reply egs2/whamr/enh1/local/data.sh:73]: done

File: egs2/whamr/enh1/local/whamr_create_mixture.sh (hunk 5 above)
Comment (Emrys365) [reply egs2/whamr/enh1/local/whamr_create_mixture.sh]: done

File: egs2/TEMPLATE/enh1/README.md (hunk 4 above)
Comment (Emrys365) [reply egs2/TEMPLATE/enh1/README.md:11]: This recipe is not yet included in this PR. We will make new commits later.

File: egs2/librimix/enh1/local/data.sh (hunk 3 above)
Comment (Emrys365) [reply egs2/librimix/enh1/local/data.sh:51]: done

File: egs2/librimix/enh1/local/data.sh (hunk 2 above)
Comment (Emrys365) [reply egs2/librimix/enh1/local/data.sh]: done

File: egs2/librimix/enh1/local/data.sh (hunk 1 above)
Comment (Emrys365) [reply egs2/librimix/enh1/local/data.sh]: done

**Comment by sw005320** [comment]
LGTM!
Once the conflict is fixed, I'll merge this PR.
It would not be easy to check whether it is robustly working only with the code review, and I think we should ask a couple of volunteers to run recipes in their environments and report the issue.
I'm happy to do it.

**Comment by sw005320** [comment]
Oh sorry, now the conflict was fixed.
I'll merge it.

**Comment by sw005320** [comment]
Many many thanks, the espnet-se team!
I'll play with some recipes!

**Comment by sw005320** [comment]
I forgot to mention.
Could you add the corpus information in https://github.com/espnet/espnet/blob/master/egs2/README.md?

**Comment by Emrys365** [comment]
> It would not be easy to check whether it is robustly working only with the code review, and I think we should ask a couple of volunteers to run recipes in their environments and report the issue.
> I'm happy to do it.

Thanks! We are asking some colleagues to help check the recipes. We may have the updated results and models in a few days. It would be better if they can be validated in other environments.

**Pull request merged**